docs/architecture/adr-014-versioned-namespaces.md

```diff
@@ -0,0 +1,195 @@
+# ADR 14: Versioned Namespaces
+
+## Status
+
+Accepted
+
+## Changelog
+
+- 2023/2/14: initial draft
+
+## Context
+
+The specs-staging branch contains the current schema for a [sparse share](https://github.com/celestiaorg/celestia-app/blob/0dc7d17636efa535efca42af58229ee0c3c21261/specs/src/specs/data_structures.md#sparse-share). A brief reminder of the schema for the first blob share in a sequence:
+
+![first-sparse-share](./assets/first-sparse-share.svg)
+
+| Field | Number of bytes | Description |
+| --- | --- | --- |
+| namespace id | 8 | namespace ID of the share |
+| info byte | 1 | the first 7 bits determine the share version. The last 1 bit determines if this is the first share in a sequence |
+| sequence length | 4 | the number of bytes in the sequence encoded as a big-endian uint32 |
+| msg1 | 499 | blob1's raw data |
+
+```go
+currentBlobShare := []byte{
+  1, 2, 3, 4, 5, 6, 7, 8, // namespace ID
+  1,          // info byte (sequence start indicator = true)
+  0, 0, 0, 3, // sequence length
+  1, 2, 3, // blob data
+  // 0 padding until share is full
+ }
+```
+
+The current schema poses a challenge for the following scenarios:
+
+1. Changes to the non-interactive default rules that don't break backwards compatibility with existing namespaces [celestia-app#1282](https://github.com/celestiaorg/celestia-app/issues/1282), [celestia-app#1161](https://github.com/celestiaorg/celestia-app/pull/1161)
+    - After mainnet launch, if we want to change the non-interactive default rules but retain the previous non-interactive default rules for backwards compatibility, it isn't possible to differentiate the namespaces that want to use the old rules vs the new rules.
```

Comment:
Just to regurgitate this point for my own understanding: We want the rules that define the construction of a square (i.e. potentially the non-interactive default rules or any other algorithm) to be on a per namespace basis. Not per square (i.e. a version number that reflects the construction algorithm for the entire square - implying homogeneity for all txs in that square) and not per share (i.e. multiple shares for the same namespace could theoretically have different versions for interpreting where they should be placed in the square - this assumes that all messages are always tied to a namespace and are always ordered (ascending)).

Reply:
Yes I believe so.

> this assumes that all messages are always tied to a namespace and are always ordered

messages => blobs ? If so, I agree.

Reply:
Well messages could also be an IBC transaction or a governance vote (they still have namespaces)